modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/standardscaling/StandardScalerTrainer.java

```diff
@@ -0,0 +1,141 @@
+
+/**
+ * Trainer of the standard scaler preprocessor.
+ *
+ * @param <K> Type of a key in {@code upstream} data.
+ * @param <V> Type of a value in {@code upstream} data.
+ */
+public class StandardScalerTrainer<K, V> implements PreprocessingTrainer<K, V, Vector, Vector> {
+    /** {@inheritDoc} */
+    @Override public StandardScalerPreprocessor<K, V> fit(DatasetBuilder<K, V> datasetBuilder,
+        IgniteBiFunction<K, V, Vector> basePreprocessor) {
+
+        SumHelper sumHelper = computeSum(datasetBuilder, basePreprocessor);
+
+        int n = sumHelper.sum.length;
+        long count = sumHelper.count;
+        double[] mean = new double[n];
+        double[] sigma = new double[n];
+
+        for (int i = 0; i < n; i++) {
+            mean[i] = sumHelper.sum[i] / count;
+            double variace = (sumHelper.squaredSum[i] - Math.pow(sumHelper.sum[i], 2) / count) / count;
+            sigma[i] = Math.sqrt(variace);
+        }
+        return new StandardScalerPreprocessor<>(mean, sigma, basePreprocessor);
+    }
+
+    /** Computes sum, squared sum and row count. */
+    private SumHelper computeSum(DatasetBuilder<K, V> datasetBuilder,
+        IgniteBiFunction<K, V, Vector> basePreprocessor) {
+        try (Dataset<EmptyContext, SumHelper> dataset = datasetBuilder.build(
+            (upstream, upstreamSize) -> new EmptyContext(),
+            (upstream, upstreamSize, ctx) -> {
+                double[] sum = null;
+                double[] squaredSum = null;
+                long count = 0;
+
+                while (upstream.hasNext()) {
+                    UpstreamEntry<K, V> entity = upstream.next();
+                    Vector row = basePreprocessor.apply(entity.getKey(), entity.getValue());
+
+                    if (sum == null) {
+                        sum = new double[row.size()];
+                        squaredSum = new double[row.size()];
+                    }
+                    else {
+                        assert sum.length == row.size() : "Base preprocessor must return exactly " + sum.length
+                            + " features";
+                    }
+
+                    ++count;
+                    for (int i = 0; i < row.size(); i++) {
+                        double x = row.get(i);
+                        sum[i] += x;
+                        squaredSum[i] += x * x;
+                    }
+                }
+                return new SumHelper(sum, squaredSum, count);
+            }
+        )) {
+
+            return dataset.compute(data -> data,
+                (a, b) -> {
+                    if (a == null)
+                        return b;
+                    if (b == null)
+                        return a;
+
+                    return a.merge(b);
+                });
+        }
+        catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /** A Service class which used for sums holing. */
+    private static class SumHelper implements AutoCloseable {
```

Comment:
Can you separate SumHelper and StandardScalerTrainer (especially in context that close() method is empty)?
It seems that SumHelper class may be usefull in other code.

Reply:
I extracted `SumHelper` to `StandardScalerData`.